### ISSUE TITLE
RxJava 2 support

### ISSUE THREAD

**Issue by vanniktech** [tbruyelle/RxPermissions#71]
Now that there's a first [RC of RxJava 2](https://github.com/ReactiveX/RxJava/releases/tag/v2.0.0-RC1) I'm wondering whether you guys want to port this library to Rx2.

Happy to help out.

**Comment by tbruyelle** [comment]
Yes I think that's definitevely something RxPermissions should support.

I read some days ago the content of RxJava2 and it's a full rewrite, so I wonder how library like RxPermissions should publish new version that breaks everything. I don't want to change the package name or the artifactId. Maybe a jump to version 1.0 is enough to warn the user about the big changes ?

**Comment by vanniktech** [comment]
> I don't want to change the package name or the artifactId

That's basically the way to go. Also RxAndroid does it that way 2 branches (1.x and 2.x) with different package ids + artifact ids.

**Comment by tbruyelle** [comment]
RxAndroid is part of the ReactiveX organisation so it sounds logic it adopted the same way.

I don't think I can use the `io.reactivex` package, and I don't want to rename the package to something like `org.tbruyelle` because it doesn't mean anything. Changing artifact id to something like `RxPermissions2` is much more meaningful but that's ugly.

**Comment by vanniktech** [comment]
Well since Rx1 & Rx2 are binary incompatible and have a different package names it's super impractical to have a library that's capable of running both versions.

How about:
`com.tbruyelle.rxpermissions` becomes `com.tbruyelle.rxpermissions2`
`com.tbruyelle.rxpermissions:rxpermissions:0.7.0@aar` becomes `com.tbruyelle.rxpermissions:rxpermissions2:0.7.0@aar`

Version 1 would be in 1.x branch, Version 2 in 2.x branch. Also I would not say that this project is any different from RxAndroid. Both rely on RxJava and build on top of it.

**Comment by tarasantoshchuk** [comment]
>That's basically the way to go. Also RxAndroid does it that way 2 branches (1.x and 2.x) with different package ids + artifact ids.

I think this approach should be accepted, because:
1) RxPermissions shouldn't force users to switch to Rx2
2) We still want users of Rx1 to use updated versions of lib with bugfixes,  etc. Which is not the case, if we switch to Rx2.

**Comment by tbruyelle** [comment]
Ok for changing the package name to `com.tbruyelle.rxpermissions2`, that's the best alternative I think. 
I'm curious to see how it will be handled in [RxBinding](https://github.com/JakeWharton/RxBinding/issues/281).

**Comment by vanniktech** [comment]
Pretty sure Jake will follow his [own advice](http://jakewharton.com/java-interoperability-policy-for-major-version-updates/)

**Comment by vanniktech** [comment]
Is someone already working on porting it?

**Comment by gengjiawen** [comment]
@vanniktech Are you working on this ?

**Comment by tbruyelle** [comment]
Not me

**Comment by vanniktech** [comment]
I might find some time next week if no one else has started on it

**Comment by tbruyelle** [comment]
I won't have the time to start that in the next weeks. 

I linked issue #65 about `Single<Boolean>` usage instead of `Observable<Boolean>`, I said this will come with rxJava2. But after some experiences I found something annoying with `Single`, they don't have `filter` operator. It's logic but for the permission request workflow, it can be annoying to don't have the `filter` operator.

**Comment by vanniktech** [comment]
Yeah it makes hardly sense to filter one emitted item. Worst case `request(..).toObservable().filter(..)` can be done.

Are you willing to support both RxJava versions in parallel meaning different package name + artifact id or should I fork it and distribute the RxJava 2 part?

**Comment by tbruyelle** [comment]
As we discussed, I'm ok for a different package name, like `com.tbruyelle.rxpermissions2` and so support both versions in parallel.

I understand your request to distribute your work on one of your repository, and if you really want to do so that's ok. But I'm not sure the community will gain from that. There will be 2 different repositories for the same project, it will be confusing. As you like.

**Comment by vanniktech** [comment]
Can you create a corresponding 2.x branch? I've got a first version ready

**Comment by tbruyelle** [comment]
Wow you're quick, thanks!
2.x branch is ready.

**Comment by vanniktech** [comment]
Created a PR: #80 - happy to receive some feedback

**Comment by gengjiawen** [comment]
@tbruyelle It seems that the rxjava2 version are not available in the jcenter.
```
* What went wrong:
A problem occurred configuring project ':app'.
> A problem occurred configuring project ':commonLibrary'.
   > Could not resolve all dependencies for configuration ':commonLibrary:_debugPublishCopy'.
      > Could not find com.tbruyelle.rxpermissions2:rxpermissions:0.7.1.
```

**Comment by tbruyelle** [comment]
@gengjiawen That's right, I don't understand why, the files are here... https://dl.bintray.com/tbruyelle/tbruyelle/com/tbruyelle/rxpermissions2/rxpermissions/0.7.1/

**Comment by tbruyelle** [comment]
Maybe it's because it's a different package ...

**Comment by gengjiawen** [comment]
The file not in here https://jcenter.bintray.com/com/tbruyelle/rxpermissions2/rxpermissions/0.7.1/rxpermissions-0.7.1.pom. Gradle will look for this url.

**Comment by tbruyelle** [comment]
Strange, I asked to bintray to understand, I'm waiting for their response.

**Comment by tbruyelle** [comment]
My config is OK because I manage to upload the legacy version at the same time.
https://jcenter.bintray.com/com/tbruyelle/rxpermissions/rxpermissions/0.7.1/rxpermissions-0.7.1.pom

**Comment by tbruyelle** [comment]
Got my answer from bintray, I have to create an other package and ask for jcenter link. Stay tuned

**Comment by tbruyelle** [comment]
@gengjiawen Request approved by bintray, please retry (I can't be myself for the moment)

**Comment by gengjiawen** [comment]
It worked :)

**Comment by gengjiawen** [comment]
@tbruyelle @vanniktech 
It's seems that rxpermission not compatibale with Rxjava2 rc4.
```
java.lang.NoSuchMethodError: No virtual method compose(Lio/reactivex/functions/Function;)
Lio/reactivex/Observable; in class Lio/reactivex/Observable; or its super classes (declaration of 'io.reactivex.Observable' appears in /data/app/com.shundaojia.travel.driver-2/base.apk)
                                                                                at com.tbruyelle.rxpermissions2.RxPermissions.request(RxPermissions.java:125)
```

**Comment by vanniktech** [comment]
I know. ObservableTransformer got changed.

**Comment by gengjiawen** [comment]
@vanniktech get time to send a pr?

**Comment by vanniktech** [comment]
https://github.com/vanniktech/RxPermissions/pull/2 although generics are still failing in tests

**Comment by gengjiawen** [comment]
Any progress for compatibility issue?

**Comment by tbruyelle** [comment]
Fixed now with version 0.8.1
